doc/sphinx/proof-engine/ltac2.rst

```diff
@@ -659,168 +734,476 @@ insert in a concise way an Ltac2 variable of type :n:`constr` into a Coq term.
-this features has the same semantics as in Ltac1. In particular, a ``reverse``
-flag can be specified to match hypotheses from the more recently introduced to
-the least recently introduced one.
+   :n:`{*, @gmatch_pattern }`
+      Patterns to match with hypotheses.  Each pattern must match a distinct hypothesis in order
+      for the branch to match.
 
-.. _ltac2_notations:
+      Hypotheses have the form :n:`@name {? := @term__binder } : @type`.  Currently Ltac2 doesn't
+      allow matching on or capturing the value of :n:`@term__binder`.  It only supports matching on
+      the :token:`name` and the :token:`type`, for example `n : ?t`.
 
-Notations
----------
+      .. currently only supports the first row
+         :list-table::
+         :widths: 2 1
+         :header-rows: 1
 
-Notations are the crux of the usability of Ltac1. We should be able to recover
-a feeling similar to the old implementation by using and abusing notations.
+         * - Pattern syntax
+           - Example pattern
 
-Scopes
-~~~~~~
+         * - :n:`@name : @ltac2_match_pattern`
+           - `n : ?t`
 
-A scope is a name given to a grammar entry used to produce some Ltac2 expression
-at parsing time. Scopes are described using a form of S-expression.
+         * - :n:`@name := @match_pattern__binder`
+           - `n := ?b`
 
-.. prodn::
-   ltac2_scope ::= {| @string | @int | @lident ({+, @ltac2_scope}) }
+         * - :n:`@name := @term__binder : @type`
+           - `n := ?b : ?t`
 
-A few scopes contain antiquotation features. For the sake of uniformity, all
-antiquotations are introduced by the syntax :n:`$@lident`.
+         * - :n:`@name := [ @match_pattern__binder ] : @ltac2_match_pattern`
+           - `n := [ ?b ] : ?t`
 
-The following scopes are built-in.
+         :token:`name` can't have a `?`.  Note that the last two forms are equivalent except that:
 
-- :n:`constr`:
+         - if the `:` in the third form has been bound to something else in a notation, you must use the fourth form.
+           Note that cmd:`Require Import` `ssreflect` loads a notation that does this.
+         - a :n:`@term__binder` such as `[ ?l ]` (e.g., denoting a singleton list after
+           :cmd:`Import` `ListNotations`) must be parenthesized or, for the fourth form,
+           use double brackets: `[ [ ?l ] ]`.
 
-  + parses :n:`c = @term` and produces :n:`constr:(c)`
+      If there are multiple :token:`gmatch_hyp_pattern`\s in a branch, there may be multiple ways to match them to hypotheses.
+      For :tacn:`match! goal` and :tacn:`multi_match! goal`, if the evaluation of the :token:`ltac2_expr` fails,
+      matching will continue with the next hypothesis combination.  When those are exhausted,
+      the next alternative from any `context` constructs in the :token:`ltac2_match_pattern`\s is tried and then,
+      when the context alternatives are exhausted, the next branch is tried.
```

Comment:
This confuses me a bit.  What if there are `context`s in the hypothesis types?  Will it try other combinations of hypotheses before trying the first combination with a different set of `context` matches?  That seems like very, very strange semantics.  (If this is only talking about the `context` constructs in goal, that is not clear in this text.)

Reply:
This is basically the same description as for Ltac1.  Submit an issue if you like.  Here's an Ltac1 example:

```
Goal forall A B : Prop, A -> ((~ True) <> (~ False)) -> (A->B).
intros A B H.
match goal with
| H1 : _, H2 : _ |- context [ (~ ?t) ] => idtac "match " H1 H2 t; fail
| _ => idtac
end.

match  B H ((~ True) = (~ False))
match  A H ((~ True) = (~ False))
match  H B ((~ True) = (~ False))
match  A B ((~ True) = (~ False))
match  H A ((~ True) = (~ False))
match  B A ((~ True) = (~ False))
match  B H True
match  A H True
match  H B True
match  A B True
match  H A True
match  B A True
match  B H False
match  A H False
match  H B False
match  A B False
match  H A False
match  B A False
```